vms/proposervm/state_syncable_vm_test.go

```diff
@@ -590,3 +594,136 @@ func TestStateSummaryAcceptOlderBlock(t *testing.T) {
 	require.NoError(err)
 	require.Equal(proBlk.ID(), lastAcceptedID)
 }
+
+// TestStateSummaryAcceptOlderBlockSkipStateSync tests the case where we accept
+// a state summary older than the last accepted block. In this case, we should not
+// roll the ProposerVM back to match the state summary, but we should invoke the
+// innerVM to accept the state summary and re-align the ProposerVM with the innerVM
+// during the transition out of state sync.
+func TestStateSummaryAcceptOlderBlockSkipStateSync(t *testing.T) {
+	require := require.New(t)
+
+	innerVM, vm := helperBuildStateSyncTestObjects(t)
+	defer func() {
+		require.NoError(vm.Shutdown(context.Background()))
+	}()
+
+	// store post fork block associated with summary
+	innerBlk1 := &snowmantest.Block{
+		Decidable: snowtest.Decidable{
+			IDV: ids.GenerateTestID(),
+		},
+		BytesV:  []byte{1},
+		ParentV: ids.GenerateTestID(),
+		HeightV: 1969,
+	}
+	innerBlk2 := snowmantest.BuildChild(innerBlk1)
+
+	innerSummary1 := &blocktest.StateSummary{
+		IDV:     innerBlk1.ID(),
+		HeightV: innerBlk1.Height(),
+		BytesV:  innerBlk1.BytesV,
+	}
+
+	require.NoError(vm.SetForkHeight(innerSummary1.Height() - 1))
+
+	// Set the last accepted block height to be higher than the state summary
+	// we are going to attempt to accept
+	vm.lastAcceptedHeight = innerBlk2.Height()
+
+	innerVM.LastAcceptedF = func(context.Context) (ids.ID, error) {
+		return innerBlk2.IDV, nil
+	}
+
+	innerVM.GetBlockF = func(_ context.Context, blkID ids.ID) (snowman.Block, error) {
+		switch blkID {
+		case snowmantest.GenesisID:
+			return snowmantest.Genesis, nil
+		case innerBlk1.ID():
+			return innerBlk1, nil
+		case innerBlk2.ID():
+			return innerBlk2, nil
+		default:
+			return nil, database.ErrNotFound
+		}
+	}
+	innerVM.GetStateSummaryF = func(context.Context, uint64) (block.StateSummary, error) {
+		return innerSummary1, nil
+	}
+	innerVM.ParseBlockF = func(_ context.Context, b []byte) (snowman.Block, error) {
+		switch {
+		case bytes.Equal(b, innerBlk1.BytesV):
+			return innerBlk1, nil
+		case bytes.Equal(b, innerBlk2.BytesV):
+			return innerBlk2, nil
+		default:
+			require.FailNow("unexpected parse block")
```

Comment:
I originally had in mind returning an error here instead of using `FailNow` to avoid the unreachable code warning... does that not work beacuse of the way this is called for some reason?

Reply:
This uses `require.FailNow(...)` because it's unexpected for this to ever parse an invalid block.

The way these tests are written we don't have an actual serialization format we just hardcode bytes and then set `ParseBlock` to match that to the block we hardcoded it to (ie. `bytes.Equal(b, innerBlk2.BytesV)`). If we see bytes that we don't recognize here, we could return an error, but it would be hiding some code path that should never occur.